readme.md
# Ipfs

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

Interact with IPFS using the [RPC api](https://docs.ipfs.tech/reference/kubo/rpc/).

Take a look at [contributing.md](contributing.md) to see a to do list.

## Installation

Via Composer

``` bash
composer require fliq/ipfs
```

## Usage

To get started with IPFS visit the [IPFS documentation](https://docs.ipfs.tech/how-to/command-line-quick-start)

This package uses [Guzzle](https://github.com/guzzle/guzzle) every request is made using the `requestAsync` method, and
every method on the Ipfs class
returns a guzzles [PromiseInterface](https://docs.guzzlephp.org/en/stable/quickstart.html#async-requests).

There are two modes for the ipfs client 'api' and 'gateway'. Gateway is accessible publicly for read only operations
and api can write.

```php
use Fliq\Ipfs\Ipfs;

$ipfs = new Ipfs('localhost', '8080', 'http', 'gateway');

$data = $ipfs->cat('QmSgvgwxZGaBLqkGyWemEDqikCqU52XxsYLKtdy3vGZ8uq')->wait(); // spaceship-launch.jpg

```

To write files you must be in 'api' mode.

```php
use Fliq\Ipfs\Ipfs;

$ipfs = new Ipfs('localhost', '5001', 'http', 'api');

$result = $ipfs->add('Hello, IPFS')->wait();

$cid = $result[0]['Hash']; // content identifier
```

You can add files flexibly to IPFS with support for strings, php resources and encode arrays into json, and multiple
files at once all wrapped into a directory.

```php
use GuzzleHttp\Psr7\Utils;

$ipfs = new Ipfs('localhost', '5001', 'http', 'api');

$file = Utils::tryFopen('path/to/file.jpeg')

$results = $ipfs->add([
    'file.jpeg' => $file,
    'meta.json' => [
        'name' => 'My file',
        'description' => 'file description',
        'properties' => [...],
    ],
], ['wrap-with-directory' => true])->wait();

$directoryCid = $results[2]['Hash'];
```

Retrieving data from IPFS. With the cat method you can get a Psr7 stream as a result, but for convenience you can
use the `get()` and `json()` methods to read files also.

```php
$data = $ipfs->cat('Qm...')
               ->then(fn($stream) => $stream->getContents())
               ->then(fn($str) => json_decode($str, 1))
               ->wait();

// is the same as.

$data = $ipfs->get('Qm...')
             ->then(fn($str) => json_decode($str, 1))
             ->wait();

// is the same as

$data = $ipfs->json('Qm...')->wait();

```

Because every request is async you can make multiple requests at the same time.

```php
use GuzzleHttp\Promise;

$promises = [];

$promises[] = $ipfs->add('Hello, IPFS');
$promises[] = $ipfs->cat('QmSgvgwxZGaBLqkGyWemEDqikCqU52XxsYLKtdy3vGZ8uq');

$responses = Promise\Utils::unwrap($promises);
```

Lastly you can call any endpoint using the `call()` method.

```php
$ipfs->call('name/resolve', ['query' => $args])->then(function(Response $response) {
    // handle response.
});
```

## Methods:

### version()
Gets Kubo node information.

### add()
Adds files to IPFS 

### cat()
Reads files returns a Stream 

#### get()
Reads files returns a string

#### json()
Reads files returns an array

### ls()
Reads content from CID

### call()
Will call 



## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

You'll need a local IPFS node running to test using localhost.

``` bash
./vendor/bin/pest
```

## Contributing

Please see [contributing.md](contributing.md) for details and a todolist.

## Security

If you discover any security related issues, please email author@email.com instead of using the issue tracker.

## Credits

- [Christian Pavilonis][link-author]
- [All Contributors][link-contributors]

## License

MIT. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/fliq/ipfs.svg?style=flat-square

[ico-downloads]: https://img.shields.io/packagist/dt/fliq/ipfs.svg?style=flat-square

[ico-travis]: https://img.shields.io/travis/fliq/ipfs/master.svg?style=flat-square

[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/fliq/ipfs

[link-downloads]: https://packagist.org/packages/fliq/ipfs

[link-travis]: https://travis-ci.org/fliq/ipfs

[link-styleci]: https://styleci.io/repos/12345678

[link-author]: https://github.com/ChristianPavilonis

[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `Ipfs` will be documented in this file.

## Version 1.0

### Added
- Everything


tests/Pest.php
<?php

//uses(TestCase::class)->in('./');


use Fliq\Ipfs\Ipfs;
use GuzzleHttp\Client;

function api() : Ipfs
{
    return new Ipfs(mode: 'api');
}

function gateway() : Ipfs
{
    return new Ipfs();
}

tests/LsTest.php
<?php
// LsTest

use Fliq\Ipfs\Commands\Ls;
use Fliq\Ipfs\Ipfs;


it('lists files', function () {
    $response = api()->add(
        ['foo.txt' => 'foo'],
        ['wrap-with-directory' => true]
    )->wait()[1];

    $command = new Ls(gateway());

    $list = $command->handle($response['Hash'])->wait();

    expect($list['Objects'][0]['Links'][0]['Name'])->toEqual('foo.txt');
});

tests/AddTest.php
<?php
// AddTest

use Fliq\Ipfs\Commands\Add;
use Fliq\Ipfs\Ipfs;
use GuzzleHttp\Psr7\Utils;

beforeEach(function () {
    $this->ipfs = api();
});

it('adds from a string', function () {
    $command = new Add(api());

    $response = $command->handle('Hello, IPFS!')->wait()[0];

    expect($response)->toHaveKey('Hash');

    $file = $this->ipfs->cat($response['Hash'])->wait();

    expect($file->getContents())->toEqual('Hello, IPFS!');
});

it('adds multiple files', function () {
    $command = new Add(api());

    $response = $command->handle('Hello, IPFS!', 'Another one!')->wait();

    expect($response)->toHaveCount(2);
});

it('wraps in a directory', function () {
    $command = new Add(api(), ['wrap-with-directory' => true]);

    $response = $command->handle('Hello, IPFS!', 'Another one!')->wait();

    expect($response)->toHaveCount(3);
});

it('can upload from a resource', function () {
    $command = new Add(api());

    $resource = Utils::tryFopen(__DIR__ . '/uploads/foo.txt', 'r');

    $response = $command->handle($resource)->wait();

    expect($response[0])->toHaveKey('Hash');
});

it('adds array as json', function () {
    $command = new Add(api());

    $response = $command->handle([
        'file.json' => [
            'foo' => 'bar',
        ],
    ])->wait();

    expect($response[0]['Name'])->toEqual('file.json');
});

src/Commands/Add.php
<?php

namespace Fliq\Ipfs\Commands;

use Fliq\Ipfs\Ipfs;
use GuzzleHttp\Client;
use GuzzleHttp\Promise\PromiseInterface;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;

class Add
{

    public function __construct(protected Ipfs $client, array $options = [])
    {
        $this->options = array_merge([
            'cid-version' => '1',
        ], $options);
    }

    public function handle(...$resources) : PromiseInterface
    {
        $promise = $this->client->call('add', [
            'multipart' => $this->build($resources),
            'query'     => $this->options,
        ]);

        return $promise->then(function (Response $response) {
            $arr = preg_split('/\n/', $response->getBody()->getContents());

            $filtered = array_filter($arr, fn($item) => trim($item) !== "");

            return array_map(fn($item) => json_decode($item, JSON_OBJECT_AS_ARRAY), $filtered);
        });
    }

    protected function build($resources) : array
    {
        $headers = [];
        $parts   = [];

        if (is_array($resources)) {
            foreach ($resources as $key => $resource) {
                // recursively resolve part if we have a numeric key.
                if (is_numeric($key)) {
                    $parts = array_merge($parts, $this->build($resource));
                }
                else { // otherwise we have a string as a key such as 'file.json' => [],
                    $headers  = $this->makeHeaders($key);

                    $parts [] = $this->makePart($headers, $resource);
                }
            }
        }
        else {
            $parts [] = $this->makePart($headers, $resources);
        }

        return $parts;
    }

    protected function makePart(array $headers, $resource) : array
    {
        if (is_array($resource)) {
            $resource = json_encode($resource);
        }

        return [
            'name'     => 'path',
            'headers'  => $headers,
            'contents' => $resource,
        ];
    }

    protected function makeHeaders(string $key) : array
    {
        return [
            'Content-Disposition' => 'form-data; name="file"; filename="' . $key . '"',
            'Content-Type'        => 'application/octet-stream',
        ];
    }

}

src/Ipfs.php
<?php

namespace Fliq\Ipfs;

use Fliq\Ipfs\Commands\Add;
use Fliq\Ipfs\Commands\Cat;
use Fliq\Ipfs\Commands\Ls;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Promise;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Psr7\Stream;
use GuzzleHttp\Psr7\Utils;
use Illuminate\Support\Facades\Http;

class Ipfs
{
    protected Client $client;
    protected string $method;

    public function __construct(
        string $host = 'localhost',
        string $port = null,
        string $protocol = 'http',
        public string $mode = 'gateway',
    ) {
        $this->method = match ($this->mode) {
            'api'     => 'POST',
            'gateway' => 'GET',
            default   => throw new IpfsApiException('mode must be api or gateway'),
        };

        $port = $port ?? match ($this->mode) {
                'api'     => '5001',
                'gateway' => '8080',
            };

        $this->client = new Client([
            'base_uri' => "{$protocol}://{$host}:{$port}/api/v0/"
        ]);
    }

    public function version() : Promise\PromiseInterface
    {
        return $this->client->requestAsync($this->method, 'version')->then(function ($response) {
            return json_decode($response->getBody()->getContents(), JSON_OBJECT_AS_ARRAY);
        });
    }

    /**
     * Adds files to IPFS.
     *
     * Can upload multiple files use array keys as file names.
     * arrays will be json encoded and uploaded as a string.
     * example:
     *
     * $ipfs->add(['file.json' => ['key' => 'value']]);
     *
     * reference: https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-add
     *
     * @param string|array|resource $resource
     * @param array $options
     *
     * @throws IpfsApiException
     */
    public function add($resource, array $options = []) : Promise\PromiseInterface
    {
        if ($this->mode != 'api') {
            throw new IpfsApiException('Must be in API mode to add files');
        }

        $command = new Add($this, $options);

        return $command->handle($resource);
    }

    /**
     * Gets the contents from the file and returns a Psr7\Stream
     * @param  mixed  $args
     * @return Promise\PromiseInterface
     */
    public function cat(mixed $args) : Promise\PromiseInterface
    {
        $command = new Cat($this);

        return $command->handle($args);
    }

    /**
     * Gets the contents of the file as a string
     * @param  mixed  $args
     * @return Promise\PromiseInterface
     */
    public function get(mixed $args) : Promise\PromiseInterface
    {
        return $this->cat($args)->then(function (Stream $response) {
            return $response->getContents();
        });
    }

    /**
     * Gets the contents of the file and json decodes it
     * @param  mixed  $args
     * @return Promise\PromiseInterface
     */
    public function json(mixed $args) : Promise\PromiseInterface
    {
        return $this->get($args)->then(function (string $contents) {
            return json_decode($contents, JSON_OBJECT_AS_ARRAY);
        });
    }

    public function ls(mixed $args) : Promise\PromiseInterface
    {

        $command = new Ls($this);
        return $command->handle($args);
    }

    /**
     * @param  string  $uri
     * @param  array  $options
     * @return Promise\PromiseInterface
     */
    public function call(string $uri, array $options = []) : Promise\PromiseInterface
    {
        return $this->client->requestAsync($this->method, $uri, $options);
    }
}